### PR TITLE
provider/gce: Don't base the name of firewall rules on hash of CIDRs

### PR DESCRIPTION
**Pull request by babbageclunk**
## Description of change

Found when testing #7234 - the updated addresses would be logged for updating the firewall, but the update wouldn't happen. The error on updating was being swallowed somewhere so it wasn't clear what was happening from the logs. (I'll fix that in a separate PR when I track it down.)

We were generating the name of the firewall using a hash of the new rules, and trying to update the firewall to have the new name. GCE will prevent a mismatch between the URL and the name with this
error:

```
opening port(s) [3306/tcp from 35.187.146.143/32]: googleapi: Error 400: Invalid value for field 'resource.name': 'juju-893de4-0-d4bc50'. The name of the resource in request body must be the same as the resource name in the URL, invalid
```
Since we can't update the name of the firewall rules it seems like the options are:

1. Create a new one (with the name still based on the hashed CIDRs) and then remove the old one.
2. Change the naming to not be based on hashing the CIDRs - use a random unique string instead.

Option 1 can't be done atomically, and cleaning up the mess if the create happened but the delete didn't seems like it would be nasty. So this change implements option 2.

Since the names will be unpredictable we need to hang on to them when we get the existing rules from GCE so we can send them back with updates/deletes. The old version of the code wasn't doing that - names were thrown away when the compute.Firewall is converted into []network.IngressRule because it could regenerate the name from the CIDRs.

## QA steps

* Bootstrap on GCE, create two models with a cross-model relation between mysql and wordpress. 
* Expose wordpress
* Check that you can successfully see the wordpress setup page at the wordpress public address.
* In the GCE console, stop the wordpress machine and then start it again - note that its public address has changed.
* Wait for the Juju instance poller to notice the updated public address.
* Check that the wordpress setup page is available at the new address.

### CODE REVIEW

**Comment by babbageclunk** [comment]
!!chittychitty!!

**Comment by jameinel** [comment]
Does this lead to us not being able to lookup a known firewall rule? Is
there a way we can have a rule that isn't 'content-based-addressing' ?

John
=:->

**Comment by babbageclunk** [comment]
@jameinel: yeah, it does cause that problem, @wallyworld pointed it out to me in IRC. Since we can't update the name of the firewall rules it seems like the options are:

1. Create a new one (with the name still based on the hashed CIDRs) and then remove the old one. 
2. Change the naming to not be based on hashing the CIDRs - use a random unique string instead.

Option 1 can't be done atomically, and cleaning up the mess if the create happened but the delete didn't seems like it would be nasty. 

So I'm in the middle of doing 2 at the moment. 

Since the names will be unpredictable we need to hang on to them when we get the existing rules from GCE so we can send them back with updates/deletes. The old version of the code wasn't doing that - names were thrown away when the `compute.Firewall` is converted into `[]network.IngressRule` because it could regenerate the name from the CIDRs.

(Sorry, I should have marked this as a WIP - I'm still fixing up tests.)

**Comment by babbageclunk** [comment]
!!chittychitty!!

File: provider/gce/google/conn_network.go [hunk 1]
```diff
@@ -13,93 +15,93 @@ import (
...
+// ingress rules. If a rule matching a set of source ranges doesn't
+// already exist, it will be created - the name will be made unique
+// using a random suffix.
 func (gce Connection) OpenPorts(target string, rules ...network.IngressRule) error {
+	return errors.Trace(gce.OpenPortsWithNamer(target, RandomSuffixNamer, rules...))
+}
+
+// FirewallNamer generates a unique name for a firewall given the firewall, a
+// prefix and a set of current firewall rule names.
+type FirewallNamer func(*firewall, string, set.Strings) (string, error)
```
Comment (wallyworld) [inline provider/gce/google/conn_network.go]: Best to name the function params to make parsing the description easier.

File: provider/gce/google/conn_network.go [hunk 2]
```diff
@@ -111,45 +113,41 @@ func (gce Connection) OpenPorts(target string, rules ...network.IngressRule) err
...
-// matchProtocolPorts returns the matching firewall name and true if any of the current
-// firewall ports (ignoring CIDR values) matches the ports to match.
-func matchProtocolPorts(currentPorts map[string]protocolPorts, match protocolPorts) (string, bool) {
-	for fwname, current := range currentPorts {
-		if current.String() == match.String() {
-			return fwname, true
-		}
+// RandomSuffixNamer tries to find a unique name for the firewall by
+// appending a random suffix.
+func RandomSuffixNamer(fw *firewall, prefix string, names set.Strings) (string, error) {
```
Comment (wallyworld) [inline provider/gce/google/conn_network.go]: could we call "names" "existingNames" for clarity

File: provider/gce/google/conn_network.go (hunk 1 above)
Comment (babbageclunk) [inline provider/gce/google/conn_network.go]: Ok

File: provider/gce/google/conn_network.go (hunk 2 above)
Comment (babbageclunk) [inline provider/gce/google/conn_network.go]: Yup

**Comment by babbageclunk** [comment]
$$merge$$

**Comment by babbageclunk** [comment]
These are spurious failures - one's in a new VSphere test that Andrew's looking at, the other is one I've seen before.

$$merge$$

**Comment by babbageclunk** [comment]
grr
$$merge$$

**Pull request merged**